Context scoping

MDCox opened the pull request:
This PR is contains work porting over the Context class from https://github.com/Shopify/liquid/blob/3146d5c3f26ca5f6f8cf58efd5562a5bb0b3c57b/lib/liquid/context.rb
and it's corresponding tests.

Part of issue https://github.com/hownowstephen/go-liquid/issues/3

The Context class handles _alot_, and at the time of this PR, these commits only add proper scoping and variable lookup.  I plan on adding more, but because full functionality requires the Strainer to be implemented I'll be stopping to implement that.  This scope of work is functional right now, however so if you want to merge feel free and I'll open a new PR for later work, otherwise I'll return and append to this branch after the Strainer is implemented.

mdcox-sdm committed:
Can add and get Vars from Context

mdcox-sdm committed:
wip commit: remove commented out ruby

Sorry, it seems like that's how you've been working on it but it's
incredibly distracting, since things could be hidden in there and I had
to search around for it.  After looking things over I should be able to
get a reasonably complete impl done.

mdcox-sdm committed:
Add scoping

mdcox-sdm committed:
Remove commented out tests

mdcox-sdm committed:
test hyphenated variable

mdcox-sdm committed:
Lookup var can find things in outer scopes!

mdcox-sdm committed:
Add test to verify scoping only goes one way

hownowstephen commented on context.go:
```diff
@@ -32,290 +65,45 @@ func (c *Context) FindVariable(e Expression) (Expression, error) {
+		return nil, err
+	}
+
+	if value, ok := scope[key]; ok {
```
Shouldn't we be calling c.Get() here?

MDCox replied:
Small semantic difference: Get() will return `nil` if the variable doesn't exist, whereas the current impl will return an ErrNotFound error.  Since it's technically valid to store and return a nil or operate on a nil it's kind of a nuanced decision what we want to do here.  I opted to leave the current behavior intact incase you'd built anything ontop of it.

I imagine it will be changed pretty soon as the pieces come together, especially error handling, but I don't know exactly what it'll look like.  If you'd rather it changed that's cool, but I'm tempted to lack inertia/need decide when to change it.

MDCox replied:
Erm, sorry I jumped the gun.  There is a difference in how it's treated, but it's probably not the best way to handle it. Incoming commit to make Get work the same way and propagate the error up, which will make it so the caller can handle it (and make judgement calls on whether it's strict or lax handling).

hownowstephen commented on context.go:
```diff
@@ -32,290 +65,45 @@ func (c *Context) FindVariable(e Expression) (Expression, error) {
+func (s *scopeStack) pop() error {
+	l := len(*s)
+	if l < 1 {
+		return errors.New(`No scopes to pop`)
```
tend to prefer factoring out into something like `ErrNoScopes = errors.New("No scopes to pop")` since it gets used in a few places

MDCox replied:
Totally right.  Makes it easier to check the Err outside the package too if it's ever raised.

hownowstephen reviewed (commented):
Looking great, thanks @MDCox! Requesting one small style change and one question about the FindVariable implementation, but otherwise ready to get it merged in 👍

mdcox-sdm committed:
Resist urge to create Err360NoScope

mdcox-sdm committed:
Change behavior of Get() to propogate err for undefined vars

Liquid allows calling undefined variables and getting `nil` in the
reference implementation.  Originally Get() followed this rule and
returned `nil` if the key was not previously set.  Instead, this
propogates a ErrVarNotFound error instead.  This allows the caller to
decide what they want to do.

This breaks 1:1 method compatibility with the Ruby impl.  But it leads
to a cleaner API design Go-wise.
